cluster-provision/gocli/pkg/k8s/k8s.go

```diff
@@ -0,0 +1,222 @@
+	scheme.AddToScheme(s)
+
+	apiextensionsv1.AddToScheme(s)
+	cephv1.AddToScheme(s)
+	monitoringv1alpha1.AddToScheme(s)
+	monitoringv1.AddToScheme(s)
+	istiov1alpha1.AddToScheme(s)
+	admissionv1.AddToScheme(s)
+	cdiv1beta1.AddToScheme(s)
+	aaqv1alpha1.AddToScheme(s)
+
+	return &K8sDynamicClientImpl{
+		client: dynamicClient,
+		scheme: s,
+	}, nil
+}
+
+func NewTestClient(reactors ...ReactorConfig) *K8sDynamicClientImpl {
+	s := runtime.NewScheme()
+	scheme.AddToScheme(s)
+
+	apiextensionsv1.AddToScheme(s)
+	cephv1.AddToScheme(s)
+	monitoringv1alpha1.AddToScheme(s)
+	monitoringv1.AddToScheme(s)
+	istiov1alpha1.AddToScheme(s)
+	admissionv1.AddToScheme(s)
+	cdiv1beta1.AddToScheme(s)
+	aaqv1alpha1.AddToScheme(s)
+
+	dynamicClient := fake.NewSimpleDynamicClient(s)
+	for _, r := range reactors {
+		dynamicClient.PrependReactor(r.verb, r.resource, r.reactfunc)
+	}
+
+	return &K8sDynamicClientImpl{
+		client: dynamicClient,
+		scheme: s,
+	}
+}
+
+func NewReactorConfig(v string, r string, reactfunc func(action testing.Action) (bool, runtime.Object, error)) ReactorConfig {
+	return ReactorConfig{
+		verb:      v,
+		resource:  r,
+		reactfunc: reactfunc,
+	}
+}
+
+func (c *K8sDynamicClientImpl) Get(gvk schema.GroupVersionKind, name, ns string) (*unstructured.Unstructured, error) {
+	resourceClient, err := c.initResourceClientForGVKAndNamespace(gvk, ns)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := resourceClient.Get(context.TODO(), name, v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+func (c *K8sDynamicClientImpl) List(gvk schema.GroupVersionKind, ns string) (*unstructured.UnstructuredList, error) {
+	resourceClient, err := c.initResourceClientForGVKAndNamespace(gvk, ns)
+	if err != nil {
+		return nil, err
+	}
+
+	objs, err := resourceClient.List(context.TODO(), v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return objs, nil
+}
+
+func (c *K8sDynamicClientImpl) Apply(manifest []byte) error {
+	yamlDocs := bytes.Split(manifest, []byte("---\n"))
+	for _, yamlDoc := range yamlDocs {
+		if len(yamlDoc) == 0 {
+			continue
+		}
+
+		jsonData, err := yaml.YAMLToJSON(yamlDoc)
+		if err != nil {
+			fmt.Printf("Error converting YAML to JSON: %v\n", err)
+			continue
+		}
+
+		obj := &unstructured.Unstructured{}
+		dec := serializer.NewCodecFactory(c.scheme).UniversalDeserializer()
+		_, _, err = dec.Decode(jsonData, nil, obj)
+		if err != nil {
+			fmt.Printf("Error decoding JSON to Unstructured object: %v\n", err)
+			continue
+		}
+
+		gvk := obj.GroupVersionKind()
+		resourceClient, err := c.initResourceClientForGVKAndNamespace(gvk, obj.GetNamespace())
+		if err != nil {
+			return err
+		}
+
+		err = c.createWithRetries(resourceClient, obj)
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Object %v applied successfully", obj.GetName())
+	}
+
+	return nil
+}
+
+func (c *K8sDynamicClientImpl) createWithRetries(resourceClient dynamic.ResourceInterface, obj *unstructured.Unstructured) error {
```

Comment:
I agree with the exponential backoff and differentiating between errors

The reason why i am retrying is because my observation when testing this is sometimes the api server becomes unavailable momentarily but eventually becomes stable. so instead of failing altogether i am retrying to see if there's an actual problem or no